packages/geocoder/src/geocoders/abstract-geocoder.ts

```diff
@@ -0,0 +1,231 @@
+import { fromCoordinates } from "@conveyal/lonlat";
+
+// Prettier does not support typescript annotation
+// eslint-disable-next-line prettier/prettier
+import type { LonLatInput } from "@conveyal/lonlat"
+import type { Feature, FeatureCollection } from "geojson"
+
+type Rect = {
+  maxLat: number;
+  maxLon: number;
+  minLat: number;
+  minLon: number;
+};
+
+type Boundary = {
+  country?: string;
+  rect?: Rect;
+};
+
+type GeocoderAPI = {
+  autocomplete: (query: AutocompleteQuery) => Promise<unknown>;
+  reverse: (query: ReverseQuery) => Promise<unknown>;
+  search: (query: SearchQuery) => Promise<unknown>;
+};
+
+// These settings get added to the request to the geocoder
+export type GeocoderConfig = {
+  apiKey?: string;
+  baseUrl?: string;
+  boundary?: Boundary;
+  focusPoint?: LonLatInput;
+  options?: RequestInit;
+  sources?: string;
+  size?: number;
+  reverseUseFeatureCollection?: boolean;
+};
+
+export type ReverseQuery = {
+  apiKey?: string;
+  format?: boolean;
+  options?: RequestInit;
+  point?: LonLatInput;
+  url?: string;
+  sources?: string;
+};
+
+export type AutocompleteQuery = {
+  apiKey?: string;
+  boundary?: Boundary;
+  focusPoint?: LonLatInput;
+  format?: boolean;
+  options?: RequestInit;
+  size?: number;
+  text?: string;
+  url?: string;
+  sources?: string;
+};
+
+export type SearchQuery = {
+  apiKey?: string;
+  boundary?: Boundary;
+  focusPoint?: LonLatInput;
+  format?: boolean;
+  options?: RequestInit;
+  size?: number;
+  text?: string;
+  url?: string;
+  sources?: string;
+};
+
+export type MultiGeocoderResponse = FeatureCollection & {
+  isomorphicMapzenSearchQuery?: SearchQuery & AutocompleteQuery & ReverseQuery;
```

Comment:
Should this be a `|` instead of `&`?

Reply:
I thought so too, but no that doesn't work. A `|` creates a type that needs to conform to any of the types, so it expects fields from all of them. Seemed counterintuitive to me, but this is what ended up working.